### PR TITLE
added ability to add in a static reference to a 3rd party lib

### PR DESCRIPTION
**Pull request by pteague**
I've needed to pull in 3rd party libraries as well so I don't have to manually list them in all of the composer.json files for different projects. I can change the static-repos.json sample to a different project if you want me to, just let me know.

### CODE REVIEW

**Comment by lemoinem** [comment]
Thank you very much for your contribution.

My point is, since the static repos potentially override the repos from GitLab, I imagine, someone will forget they have set such a repo and start wondering why the updates to their repo doesn't do anything for THIS repo. Having a flag specifying that the repo is from a static source and not from GitLab will instantly help recognize this issue.

Therefore, For debugging purposes, could you, automatically add an item in the packages' `extra`s named `_source`, set to `static`, or something similar. Of course, make sure this doesn't override any extra already set by the package, please.

Once this is good, I'm ready to merge the feature in.

Thanks.

**Comment by pteague** [comment]
Did you want that in the sample static json file or programmatically added in?

**Comment by lemoinem** [comment]
No, I meant programmatically added in, it shouldn't be in the sample...

**Comment by pteague** [comment]
Ok, that should work. Only issue might be if some other project actually makes use of "_source" within the "extra" section, but I'm not currently aware of any.

File: htdocs/packages.php [hunk 1]
```diff
@@ -166,6 +167,22 @@
             $packages[$project['path_with_namespace']] = $package;
         }
     }
+    if ( file_exists( $static_file ) ) {
+        $static_packages = json_decode( file_get_contents( $static_file ) );
+        foreach ( $static_packages as $name => $package ) {
+            foreach ( $package as $version => $root ) {
+                if ( isset( $root->extra ) ) {
+                    $root->extra->_source = 'static';
```
Comment (lemoinem) [inline htdocs/packages.php]: Could you put an extra security in case `_source` already exists in the package definition?

Comment (pteague) [inline htdocs/packages.php]: So if ``_source`` already exists, leave it alone?

Comment (lemoinem) [inline htdocs/packages.php]: Yup and add an additional `_` in front of it so we can still track the static source. Lather, rinse, repeat...

File: htdocs/packages.php [hunk 2]
```diff
@@ -172,8 +172,15 @@
         foreach ( $static_packages as $name => $package ) {
             foreach ( $package as $version => $root ) {
                 if ( isset( $root->extra ) ) {
-                    if ( !isset( $root->extra->_source ) ) {
-                        $root->extra->_source = 'static';
+                    $source = '_source';
+                    while ( true ) {
```
Comment (lemoinem) [inline htdocs/packages.php]: Maybe something more along the lines of
```php
while ( isset( $root->extra->{$source} ) {
   $source = '_'.$source;
}
$root->extra->{$source} = 'static';
```

**Comment by lemoinem** [comment]
thank you for putting on with all my requests :)
Great Job, please, keep up the good work!

**Comment by pteague** [comment]
np & thanks for letting me & BigE contribute (I had to rewrite all our previous updates because of code changes before I realized we hadn't made a pull request, took so long to get permission). Last place I used this closed down, but I still use my own GitLab. Places I've been interviewing all use git, so might be introducing this again elsewhere.

**Pull request merged**